src/backend/d3d12/ResourceUploader.cpp

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017 The NXT Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ResourceUploader.h"
+
+#include "D3D12Backend.h"
+
+namespace backend {
+namespace d3d12 {
+
+    ResourceUploader::ResourceUploader(Device* device) : device(device) {
+        ASSERT_SUCCESS(device->GetD3D12Device()->CreateFence(uploadSerial, D3D12_FENCE_FLAG_NONE, IID_PPV_ARGS(&fence)));
+    }
+
+    void ResourceUploader::UploadResource(ComPtr<ID3D12Resource> resource, uint32_t start, uint32_t count, const uint32_t* data) {
+        D3D12_RESOURCE_DESC resourceDescriptor = resource->GetDesc();
```

Comment:
What do you mean by this? This is being done because there is an existing resource on the GPU which we're uploading to. The resource created in the upload heap to copy from needs to have the same descriptor.